src/interactions/slashCommands/SlashCommandOptions.ts

```diff
@@ -0,0 +1,285 @@
+	 * @param input A function that returns an option builder, or an already built builder
+	 */
+	public addRoleOption(input: SlashCommandRoleOption | ((builder: SlashCommandRoleOption) => SlashCommandRoleOption)) {
+		return this._sharedAddOptionMethod(input, SlashCommandRoleOption);
+	}
+
+	/**
+	 * Adds a mentionable option
+	 * @param input A function that returns an option builder, or an already built builder
+	 */
+	public addMentionableOption(
+		input: SlashCommandMentionableOption | ((builder: SlashCommandMentionableOption) => SlashCommandMentionableOption),
+	) {
+		return this._sharedAddOptionMethod(input, SlashCommandMentionableOption);
+	}
+
+	/**
+	 * Adds a string option
+	 * @param input A function that returns an option builder, or an already built builder
+	 */
+	public addStringOption(
+		input: SlashCommandStringOption | ((builder: SlashCommandStringOption) => SlashCommandStringOption),
+	) {
+		return this._sharedAddOptionMethod(input, SlashCommandStringOption);
+	}
+
+	/**
+	 * Adds an integer option
+	 * @param input A function that returns an option builder, or an already built builder
+	 */
+	public addIntegerOption(
+		input: SlashCommandIntegerOption | ((builder: SlashCommandIntegerOption) => SlashCommandIntegerOption),
+	) {
+		return this._sharedAddOptionMethod(input, SlashCommandIntegerOption);
+	}
+
+	private _sharedAddOptionMethod<T extends SlashCommandOptionBase>(
+		input: T | ((builder: T) => T),
+		Instance: new () => T,
+	): R {
+		const { options } = this;
+
+		// First, assert options conditions - we cannot have more than 25 options
+		validateMaxOptionsLength(options);
+
+		// Get the final result
+		const result = typeof input === 'function' ? input(new Instance()) : input;
+
+		if (!(result instanceof Instance))
+			throw new TypeError(
+				`Expected to receive a ${Instance.name} builder back, got ${
+					// eslint-disable-next-line @typescript-eslint/no-unnecessary-condition
+					((result as any)?.name as string) ?? 'Unknown'
+				} ("${typeof result}") instead`,
+			);
+
+		// Push it
+		options.push(result);
+
+		return this as any;
+	}
+}
+
+export class SharedNameAndDescription {
+	protected name!: string;
+	protected description!: string;
+
+	/**
+	 * Sets the name
+	 * @param name The name
+	 */
+	public setName(name: string) {
+		// Assert the name matches the conditions
+		validateName(name);
+
+		this.name = name;
+
+		return this;
+	}
+
+	/**
+	 * Sets the description
+	 * @param description The description
+	 */
+	public setDescription(description: string) {
+		// Assert the description matches the conditions
+		validateDescription(description);
+
+		this.description = description;
+
+		return this;
+	}
+}
+
+class SlashCommandOptionBase extends SharedNameAndDescription implements ToAPIApplicationCommandOptions {
+	protected required = false;
+	protected type: ApplicationCommandOptionType;
+
+	public constructor(type: ApplicationCommandOptionType) {
+		super();
+		this.type = type;
```

Comment:
Hear me out:

```suggestion
	protected required = false;
	protected type = this._type;

	public constructor(private _type: ApplicationCommandOptionType) {
		super();
```

Reply:
Crawl, no.

If you want to skip the second type definition, deal with it.

Reply:
![image](https://user-images.githubusercontent.com/20760160/124024952-c7f00200-d9e7-11eb-9c1f-c5b74169136a.png)

Reply:
You two are gonna give me a headache, either we leave this as is or we're making it back to `protected type: ApplicationCommandOptionType` inlined

Reply:
That's not even the syntax of the future, that's using the syntax of the past.

`??=` is from this year, whereas this syntax existed for over 4 years.